source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1027 @@
+  mostly irrelevant however, since we can't really define a list element
+  multiple times.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 24f9c34..fd5027a 100644
+--- a/default.nix
++++ b/default.nix
+@@ -4,12 +4,21 @@
+     generate.script = lib.mkOption {
+       type = lib.types.lines;
+     };
++
++    generate.requestParams = lib.mkOption {
++      type = lib.types.listOf lib.types.str;
++    };
+   };
+ 
+   config = {
+     generate.script = ''
+       map size=640x640 scale=2 | icat
+     '';
++
++    generate.requestParams = [
++      "size=640x640"
++      "scale=2"
++    ];
+   };
+   
+ }
+```
+
+## Dependencies between options: Using generate.requestParams
+
+A collection of modules generally only has a single option that is meant
+to be evaluated. That's the option that generates the final result we're
+interested in, which in our case is `generate.script`.
+
+In order to build up abstractions on that, we have the ability for
+options to depend on other options. In this case we want the
+`generate.script` option to use the values of `generate.requestParams`.
+We can access the values of options by adding the `config` argument to
+the argument list at the top and using e.g.
+`config.generate.requestParams` to access that options value.
+
+We're then using `lib.concatStringsSep " "` to join each list element
+of the option together into an argument list.
+
+```diff
+diff --git a/default.nix b/default.nix
+index fd5027a..050a98e 100644
+--- a/default.nix
++++ b/default.nix
+@@ -1,4 +1,4 @@
+-{ lib, ... }: {
++{ lib, config, ... }: {
+ 
+   options = {
+     generate.script = lib.mkOption {
+@@ -12,7 +12,9 @@
+ 
+   config = {
+     generate.script = ''
+-      map size=640x640 scale=2 | icat
++      map ${lib.concatStringsSep " "
++            config.generate.requestParams
++           } | icat
+     '';
+ 
+     generate.requestParams = [
+```
+
+## Conditional definitions: Introducing map.zoom option
+
+We now introduce the new option `map.zoom` in order to control the zoom
+level of the map. We'll use a new type for it, `nullOr <type>`, which
+accepts the value `null`, but also values of its argument type. We're
+using `null` here to mean an inferred zoom level.
+
+For this option, we'll set a default value which should be used if it's
+not defined otherwise, which we can do using `mkOption`'s `default`
+argument.
+
+Now we want to use this option to define another element in
+`generate.requestParams`, but we only want to add this element if its
+value is non-null. We can do this using the `mkIf <condition>
+<definition>` function, which only adds a definition if the condition
+holds.
+
+```diff
+diff --git a/default.nix b/default.nix
+index 050a98e..6b6e1e1 100644
+--- a/default.nix
++++ b/default.nix
+@@ -8,6 +8,13 @@
+     generate.requestParams = lib.mkOption {
+       type = lib.types.listOf lib.types.str;
+     };
++
++    map = {
++      zoom = lib.mkOption {
++        type = lib.types.nullOr lib.types.int;
```

Comment:
How can this ever be null if there's a default value?

Reply:
If it's explicitly set with `config.zoom = null;`